src/javaharness/java/arcs/api/DeviceClientImpl.java

```diff
@@ -0,0 +1,78 @@
+package arcs.api;
+
+import arcs.api.ArcsEnvironment;
+import arcs.api.DeviceClient;
+import arcs.api.PortableJson;
+import arcs.api.PortableJsonParser;
+import arcs.api.PECInnerPort;
+import arcs.api.PECInnerPortFactory;
+import arcs.api.NativeParticle;
+
+import javax.inject.Inject;
+import java.util.HashMap;
+import java.util.Map;
+
+public class DeviceClientImpl implements DeviceClient {
+    private final String FIELD_MESSAGE = "message";
+    private final String MESSAGE_READY = "ready";
+    private final String MESSAGE_DATA = "data";
+    private final String MESSAGE_PEC = "pec";
+    private final String FIELD_TRANSACTION_ID = "tid";
+    private final String FIELD_DATA = "data";
+    private final String FIELD_PEC_ID = "id";
+
+    private final PortableJsonParser jsonParser;
+    private Map<String, ArcsEnvironment.DataListener> inProgress;
+    private final PECInnerPortFactory portFactory;
+    private final Map<String, PECInnerPort> portById = new HashMap<String, PECInnerPort>();
+
+    @Inject
+    public DeviceClientImpl(PortableJsonParser jsonParser,
+                            Map<String, ArcsEnvironment.DataListener> inProgress,
+                            PECInnerPortFactory portFactory) {
+        this.jsonParser = jsonParser;
+        this.inProgress = inProgress;
+        this.portFactory = portFactory;
+    }
+
+    @Override
+    public void receive(String json) {
+        // log("receive called " + json);
+        PortableJson content = jsonParser.parse(json);
+        String message = content.getString(FIELD_MESSAGE);
+        switch (message) {
+            case MESSAGE_READY:
+                log("Received 'ready' message");
+                break;
+            case MESSAGE_DATA:
+                String transactionId = String.valueOf(content.getInt(FIELD_TRANSACTION_ID));
+                if (inProgress.containsKey(transactionId)) {
+                    PortableJson dataJson = content.getObject(FIELD_DATA);
+                    if (dataJson != null) {
+                        inProgress.get(transactionId).onData(transactionId, jsonParser.stringify(dataJson));
+                    }
+                    inProgress.remove(transactionId);
+                }
+                break;
+            case MESSAGE_PEC:
+                postMessage(content.getObject(FIELD_DATA));
+                break;
+            default:
+                throw new AssertionError("Received unsupported message: " + message);
+        }
+    }
+
+    private void postMessage(PortableJson msg) {
+        PECInnerPort port = getOrCreatePort(msg.getString(FIELD_PEC_ID));
+        port.handleMessage(msg);
+    }
+
+    private PECInnerPort getOrCreatePort(String id) {
+        if (!this.portById.containsKey(id)) {
+            this.portById.put(id, this.portFactory.createPECInnerPort(id));
+        }
+        return this.portById.get(id);
+    }
+
+    public void log(String txt) {}
```

Comment:
We could probably remove the need to have subclasses by injecting a Logger via dependency injection.

Or you could use java.util.Logger as an implementation. On Android, it could be wired up to Android Logger and on J2CL it already logs it to console. But let's leave that for later.

Reply:
Thanks! I was going to add a TODO, but it was too ugly altogether and I just switched to using java.util.logging.Logger now.